### ISSUE TITLE
Automated sample finding for robot TEM scans

### ISSUE THREAD

**Issue by CJ-Wright** [xpdAcq/mission-control#127]
Finding the TEM grid sample is difficult and time expensive, it would be great if we could have a scan find and center it.

## Expected Behavior
Run scan, centers sample and tells user how big sample is or runs the finer scan.
## Steps to Reproduce (for bugs)

1.
2.
3.

**Comment by sbillinge** [comment]
It strikes me we would like to have a more abstract "optimization" scan that takes in a set of motors and scan ranges, and a cost function, and finds the motor positions that optimizes the cost function.  The cost function could be any QOI

**Comment by dooryhee** [comment]
A similar use case is:
- robot: load sample (type 2 holder, i.e. flat plate)
- diffractometer omega (sample circle): rotate in transmission geometry
- photodiode: insert and read current from transmitted beam thru sample
- diffractometer diff_X and diff_Y: 2D scan to locate the sample in the beam (search for a dip in the photodiode current curve)
- collect data

**Comment by sbillinge** [comment]
@dooryhee I think this is the same UC.  You don't want your UC per-se, but you are proposing a way of implementing the UC you really want which is:
1. user loads flat grid sample on the diffractometer and closes the hutch
1. user starts flat-plate-well scan
1. xpdAcq automatically maps out the positions of the sample-wells
1. xpdAcq collects good data from each well, and stores the resulting data, with the position and the chemical composition etc, information.

There may be a UC2 which is:
1. same as UC1 but (1) is replaced by robot picks up the flat well-plate and loads it onto the diffractometer (or just holds it)

btw, if positioning of the plate is not reproducible we will have to do some kind of autocalibration on it...maybe by loading Ni into 3 of the wells at the corners.

**Comment by CJ-Wright** [comment]
@MehmetTopsakal
What would be a good metric for the wells? Can we go by total incident photons on the detector, or do we need something more elaborate (peak intensity, peak position)?

**Comment by dooryhee** [comment]
Either the intensity of the transmitted through beam or the total diffraction counts on the detector.

**Comment by CJ-Wright** [comment]
I think the next step is to take the code and to turn it into functions which can be composed into a pipeline.

**Comment by MehmetTopsakal** [comment]
@CJ-Wright 
I am about to finish functions.

**Comment by MehmetTopsakal** [comment]
@CJ-Wright

Suppose we have a sample with the name "AJ_XRD". After or during experiment, its tiff files, dark_sub, meta info, iq, itth... are all stored at: 

- /xxx/yyyy/AJ_XRD

I have a python class called SAMPLE. It is capable of reading all information in AJ_XRD folder.

```python
from XXX import SAMPLE

smpl = SAMPLE(sample_name='AJ_XRD',root_folder='/xxx/yyyy',load_now=False)
smpl.read() # this reads tiff and .chi files
smpl.integrate() # integrates tiff files if needed
smpl.save() # saves state as 'saved.pkl'
smpl.plot()
```
generates this plot:
![image](https://user-images.githubusercontent.com/12613640/45934803-3bc86a80-bf71-11e8-8b00-3e9355b6d60a.png)

For finding center of sample, I have center_finder function

```python
from XXX import center_finder

center = center_finder(smpl,meta_id=None,qrange=[1,5],st_ratio=0.5, use_unclassified=True,
                            dbscan_params={'eps':0.05,'min_samples':20,'n_jobs':-1},
                            plot = True, export_fig=True)
print(center)
```

returns [49.36255172 -8.28031034] as the center of sample and generates this plot:

![image](https://user-images.githubusercontent.com/12613640/45934919-ee4cfd00-bf72-11e8-9908-07752a81cc0a.png)

Star (*) denotes approximate center of sample.

I have used SAMPLE class and center_finder function for many other previously measured samples and default arguments work perfectly.


My next question is: Where should I put this SAMPLE and center_finder function for XPD integration?


Thanks
Mehmet

**Comment by MehmetTopsakal** [comment]
@CJ-Wright 

Would this function work?

```python
import numpy as np
from sklearn.cluster import DBSCAN
from scipy.stats import spearmanr

def find_center(Q_arr, I_arr, xy_arr, params):
    
    ''' 
    Trim to selected Q range.
    Because we do not want mess around sample holder and detector edge.
    '''
    sel = (Q_arr > params['qrange'][0]) & (Q_arr < params['qrange'][1])
    I_arr_sel = []
    for i in I_arr:
        I_arr_sel.append(i[sel])


    '''
    Use DBSCAN algorithm:
    - http://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html
    to cluster I_arr.
    '''
    dbs   = DBSCAN(params['eps'], min_samples=params['min_samples'],
                metric=lambda i, j: 1 - spearmanr(i, j)[0], n_jobs=params['n_jobs'])
    
    '''
    Negative prediction (-1) means DBSCAN is unable to cluster that one.
    This happens around sample boundaries. If use_unclassified=True, it is included.
    '''
    preds = dbs.fit_predict(np.array(I_arr_sel))
    
    
    uniques, counts = np.unique(preds, return_counts=True)
    ratios = counts / sum(counts)

    
    '''
    Here I collect dx,dy points for determining points which should 
    correspond to the sample.
    '''
    pts = []   

    for j,u in enumerate(uniques):
        mask = (preds == u)        
        masked = []
        for i,tf in enumerate(mask):
            if tf:
                masked.append(dxdy_arr[i])
        # these are not classified
        if u == -1:
            if params['use_unclassified']:
                pts.extend(masked)
        else:
            # these are possibly background
            if not (ratios[j] >= params['s_ratio']):
                pts.extend(masked)        
    pts = np.array(pts)
    
    # Center is calculated as the mean of the collected points.    
    center = np.mean(pts, axis=0) 

    return center, pts
```


- Q_arr is the Q points.
- xy_arr is the array of dx,dy positions
- I_arr is the array of intensities.

For 2875 Q-points and 11x11 scan points: shape of xy_arr is 121x2, I_arr is 121x2875.

Unfortunately, I have no idea how to get these arrays from bluesky stuff.

Can you help me on that?

Thanks

**Comment by MehmetTopsakal** [comment]
@CJ-Wright 

Below is how I run the function:
 

```python
'''
We need to provide same parameters for DBSCAN algorithm.
See DBSCAN documentation:
http://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html
- eps: The maximum distance between two samples for them to 
  be considered as in the same neighborhood.
- min_samples: The number of samples (or total weight) 
  in a neighborhood for a point to be considered as a core point. 
  This includes the point itself.
- n_jobs: The number of parallel jobs to run. None means 1 
  unless in a joblib.parallel_backend context. 
  -1 means using all processors.

In addition,
s_ratio: Ratio of sample points to total points. If we have
100 dx,dy points and sample is expected to be found in 20 of
these points, then the remaining 80 should be background.
So, s_ratio is 0.2. Default value of 0.5 works fine.

use_unclassified: Sometimes DBSCAN is unable to classify 
points around sample boundary. It that case, it gives -1.
If this keyword is True, that point is considered in the 
sample positions (pts).
'''
params = {'eps':0.05,'min_samples':20,'n_jobs':-1,
          's_ratio':0.5,'qrange':(1,5),
         'use_unclassified':True}


center, pts = find_center(Q_arr, I_arr, dxdy_arr, params)
print(center)
print(pts)
```


which returns

[49.38535484 -8.28809677]
[[50.966 -7.901]
 [47.966 -8.901]
 [50.966 -8.401]
 [48.466 -9.401]
 [49.966 -9.401]
 [50.966 -7.401]
 [49.466 -7.401]
 [50.466 -8.401]
 [49.966 -7.401]
 [48.966 -8.401]
 [50.466 -8.901]
 [49.466 -8.401]
 [49.966 -7.901]
 [48.966 -7.401]
 [47.966 -8.401]
 [48.466 -7.901]
 [50.466 -7.401]
 [49.466 -9.401]
 [48.966 -9.401]
 [50.466 -7.901]
 [49.966 -8.401]
 [48.466 -8.401]
 [48.466 -8.901]
 [49.466 -8.901]
 [49.466 -7.901]
 [47.966 -7.401]
 [49.966 -8.901]
 [48.966 -7.901]
 [48.966 -8.901]
 [47.966 -7.901]
 [48.466 -7.401]]

**Comment by CJ-Wright** [comment]
Yes, that looks like it will work. The pipeline can extract the xy coordinates from the live streaming data. How are the Q/I arrays formatted? Are they a list of arrays or are they 2D arrays?

Would it be possible to put this into a PR into xpdtools, in the [tools module](https://github.com/xpdAcq/xpdtools/blob/master/xpdtools/tools.py)? This looks good, I have a few minor comments, but it would be easier to make them on a PR.

**Comment by sbillinge** [comment]
@MehmetTopsakal <https://github.com/MehmetTopsakal>  let us know if need
any guidance on the group git workflow.

**Comment by sbillinge** [comment]
@MehmetTopsakal we would like to pull this in, but we need tests.  Please can you let us know if you will be able to write the tests in time to include it in the 2019 cycle 1 milestone?

**Comment by MehmetTopsakal** [comment]
@sbillinge 
Sorry about my silence on this.

This is the function:
```python
from sklearn.cluster import DBSCAN
from scipy.stats import spearmanr
def find_sample_from_2dscan(I_arr, xy_arr, Q_arr=None, 
        eps=0.05, min_samples=20, n_jobs=1,
        b_ratio_thres=0.5, qrange=(1,5), use_unclassified=True):
    """Find sample positions from xy-scan 
    
    Parameters
    ----------
    xy_arr : x,y of scan points
    I_arr : Intensities for each scan point
    Q_arr : Q-points (optional)
    
    See DBSCAN documentation:
    http://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html
    
    eps: The maximum distance between two samples for them to 
      be considered as in the same neighborhood.
      
    min_samples: The number of samples (or total weight) 
      in a neighborhood for a point to be considered as a core point. 
      This includes the point itself.
      
    n_jobs: The number of parallel jobs to run. None means 1 
      unless in a joblib.parallel_backend context. 
      -1 means using all processors.  
    
    b_ratio_thres: Clusters more than this threshold will be 
    condidered as background (not belonging to sample).
    use_unclassified: Sometimes DBSCAN is unable to classify 
    points around sample boundary. It that case, it gives -1.
    If this keyword is True, that point is considered in the 
    sample positions (pts).
    
    Returns
    -------
    center : ndarray
        xy coordinates of the center of the sample
    pts : ndarray
        xy coordinates of points considered within the sample
        
    Questions? 
    Please contact Mehmet Topsakal (mtopsakal@bnl.gov)
    
    """
    
    if isinstance(Q_arr,np.ndarray):
        # Trim to selected Q range. Because we do not want mess around 
        # beam stopper and high q. This also speedups DBSCAN calculation
        sel = (Q_arr > qrange[0]) & (Q_arr < qrange[1])            
        I_arr = np.array([i[sel] for i in I_arr])
    else:
        print('Q array is not provided. Using all intensities')
                
    # Use DBSCAN package to cluster I_arr
    dbs   = DBSCAN(eps, min_samples=min_samples,
                metric=lambda i, j: 1 - spearmanr(i, j)[0], n_jobs=n_jobs)
    preds = dbs.fit_predict(np.array(I_arr))
    uniques, counts = np.unique(preds, return_counts=True)
    ratios = counts / sum(counts)
    
    # Collect x,y data for determining points which should correspond to the sample.
    pts = []   
    for j,u in enumerate(uniques):
        mask = (preds == u)        
        masked = []
        for i,tf in enumerate(mask):
            if tf:
                masked.append(xy_arr[i])              
        if u == -1:
            if use_unclassified:
                pts.extend(masked)
        else:
            if (ratios[j] <= b_ratio_thres):
                pts.extend(masked)        
    pts = np.array(pts)
       
    center = np.mean(pts, axis=0) 

    # TODO: Get rid of b_ratio_thres
    
    return center, pts
```


here are the fake arrays:
```python
xy_arr = []
for x in np.linspace(-6,4,11):
    for y in np.linspace(3,-7,11):
        xy_arr.append([x,y])
xy_arr = np.array(xy_arr)        

# fake intensites for sample and background
I_bkg = [0.1,0.0,0.0,0.0,0.0,0.0,0.2,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.1]
I_smpl= [0.0,0.0,1.0,2.0,3.0,8.0,4.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,6.0,120.0,5.0,0.0,0.0,0.0] 

I_arr = []
for i in xy_arr:
    if np.sqrt(i[0]**2+i[1]**2) <= 3.0:
        I_arr.append(I_smpl)
    else:
        I_arr.append(I_bkg)
I_arr = np.array(I_arr)   
```

after running the function
```python
center, pts  = find_sample_from_2dscan(I_arr, xy_arr)
print(center)
```
We should get center of sample as "[0. 0.]"

![image](https://user-images.githubusercontent.com/12613640/50728485-81f8a280-10f8-11e9-9289-e572c3f8f4e8.png)


@CJ-Wright 
Can you put above function and test arrays into xpdtools for me?

Thanks

**Comment by CJ-Wright** [comment]
I think there is an existing PR for this that needs some tests. xref: https://github.com/xpdAcq/xpdtools/pull/57/files
